FerramAerospaceResearch/LEGACYferram4/FARControllableSurface.cs

```diff
@@ -547,6 +547,8 @@ public void CalculateSurfaceFunctions()
             float mass = 0;
             foreach (Part p in VesselPartList)
             {
+                if (p == null)
+                    continue;
```

Comment:
Have you ever encountered `null`s here? `p` should never be null unless KSP returns such part lists

Reply:
Yes, quite a lot of them in BDA tournaments with proc wings. This is likely due to the ordering of events being resolved since both `UpdateShipPartsList` and `CalculateSurfaceFunctions` are triggered by `OnVesselPartsChange`. Perhaps this should be calling `GetShipPartList` prior to iterating through the list instead of just skipping the parts if they're destroyed?

Reply:
Really, `UpdateShipPartsList` should be called in `ForceOnVesselPartsChange` before invoking `OnVesselPartsChange` to not rely on the invocation order.

Reply:
But `ForceOnVesselPartsChange` is called from `ForceLegacyAeroUpdates`, which is only called in `FixedUpdate` in FARVesselAero.cs, so it wouldn't be being updated in time for the event handling.

Reply:
`ForceOnVesselPartsChange` also invokes `CalculateSurfaceFunctions`. Can you try calling `UpdateShipPartsList` in `ForceOnVesselPartsChange` before the delegate to see if it changes anything?

Reply:
I haven't had any exceptions from this since adding the `UpdateShipPartsList` in `ForceOnVesselPartsChange`, so I guess that makes the `p == null` check unnecessary.

Reply:
Yeah, if rather not have have Unity Object comparisons to null if possible

Reply:
Ah, no, wait, I had the debug log level wrong in the `FARLogger` and it just wasn't printing the debug message I was looking for. `VesselPartList` does still contain `null` entries.

Reply:
In this case it might be best to remove all null parts in `UpdateShipPartsList` instead of doing null checks in multiple places

Reply:
I still think it's being caused by the ordering of event handling. I'm going to test adding a call to `UpdateShipPartsList` right before that loop to see if that helps first.

Reply:
That seemed to have worked. The last 11 rounds of my tournament ran without any exceptions.